File: src/game/furniture.ts
import { upgradesForRoom, type Upgrade } from './upgrades'

export type SceneLayout = 'desktop' | 'mobile'
export interface FurniturePoint { x: number; y: number }
export type FurniturePosition = Partial<Record<SceneLayout, FurniturePoint>>

export function visibleFurniture(roomId: number, boughtUpgrades: string[]): Upgrade[] {
  const owned = new Set(boughtUpgrades)
  const items = upgradesForRoom(roomId)
  return Array.from({ length: 5 }, (_, slot) =>
    items.find((item) => item.slot === slot && item.tier === 'advanced' && owned.has(item.id))
      ?? items.find((item) => item.slot === slot && item.tier === 'basic' && owned.has(item.id)),
  ).filter((item): item is Upgrade => Boolean(item))
}

export function defaultFurniturePosition(upgrade: Upgrade): FurniturePosition {
  const { x, y, mobileX, mobileY } = upgrade.placement
  return {
    desktop: { x, y },
    mobile: { x: Math.max(19, Math.min(81, mobileX ?? x)), y: mobileY ?? y },
  }
}

export function positionForLayout(position: FurniturePosition | undefined, layout: SceneLayout): FurniturePoint | null {
  return position?.[layout] ?? position?.[layout === 'desktop' ? 'mobile' : 'desktop'] ?? null
}

export function constrainFurniturePoint(upgrade: Upgrade, layout: SceneLayout, point: FurniturePoint): FurniturePoint | null {
  if (!Number.isFinite(point.x) || !Number.isFinite(point.y)) return null
  const width = layout === 'mobile' ? Math.min(35, upgrade.placement.width * 1.4) : upgrade.placement.width
  const height = layout === 'mobile' ? upgrade.placement.mobileHeight ?? upgrade.placement.height : upgrade.placement.height
  const edge = width / 2 + 2
  const minimumY = upgrade.surface === 'wall' ? Math.max(17, height / 2 + 4) : 57
  const maximumY = upgrade.surface === 'wall' ? Math.min(49, 55 - height / 2) : 97 - height / 2
  const clamp = (value: number, min: number, max: number) => Math.round(Math.max(min, Math.min(max, value)) * 10) / 10
  return { x: clamp(point.x, edge, 100 - edge), y: clamp(point.y, minimumY, maximumY) }
}

File: android/game/src/main/kotlin/dev/psychocat/catclicker/game/data/Furniture.kt
package dev.psychocat.catclicker.game.data

import kotlin.math.floor
import kotlin.math.max
import kotlin.math.min

/** Port of src/game/furniture.ts. Portrait phones use [SceneLayout.MOBILE], landscape and tablets [SceneLayout.DESKTOP]. */
enum class SceneLayout(val key: String) { DESKTOP("desktop"), MOBILE("mobile") }

data class FurniturePoint(val x: Double, val y: Double)

/** Positions of one item in each layout; a missing layout means "not placed yet" for that layout. */
data class FurniturePosition(val desktop: FurniturePoint? = null, val mobile: FurniturePoint? = null) {
    operator fun get(layout: SceneLayout): FurniturePoint? = if (layout == SceneLayout.DESKTOP) desktop else mobile

    fun with(layout: SceneLayout, point: FurniturePoint): FurniturePosition =
        if (layout == SceneLayout.DESKTOP) copy(desktop = point) else copy(mobile = point)
}

object Furniture {
    /** One item per slot: the advanced version replaces the basic one once bought. */
    fun visible(roomId: Int, boughtUpgrades: Collection<String>): List<Upgrade> {
        val owned = boughtUpgrades.toSet()
        val items = Upgrades.forRoom(roomId)
        return (0 until 5).mapNotNull { slot ->
            items.firstOrNull { it.slot == slot && it.tier == UpgradeTier.ADVANCED && it.id in owned }
                ?: items.firstOrNull { it.slot == slot && it.tier == UpgradeTier.BASIC && it.id in owned }
        }
    }

    fun defaultPosition(upgrade: Upgrade): FurniturePosition {
        val p = upgrade.placement
        return FurniturePosition(
            desktop = FurniturePoint(p.x, p.y),
            mobile = FurniturePoint(max(19.0, min(81.0, p.mobileX ?: p.x)), p.mobileY ?: p.y),
        )
    }

    fun positionForLayout(position: FurniturePosition?, layout: SceneLayout): FurniturePoint? {
        val other = if (layout == SceneLayout.DESKTOP) SceneLayout.MOBILE else SceneLayout.DESKTOP
        return position?.get(layout) ?: position?.get(other)
    }

    fun constrainPoint(upgrade: Upgrade, layout: SceneLayout, point: FurniturePoint): FurniturePoint? {
        if (!point.x.isFinite() || !point.y.isFinite()) return null
        val p = upgrade.placement
        val mobile = layout == SceneLayout.MOBILE
        val width = if (mobile) min(35.0, p.width * 1.4) else p.width
        val height = if (mobile) p.mobileHeight ?: p.height else p.height
        val edge = width / 2 + 2
        val wall = upgrade.surface == FurnitureSurface.WALL
        val minimumY = if (wall) max(17.0, height / 2 + 4) else 57.0
        val maximumY = if (wall) min(49.0, 55 - height / 2) else 97 - height / 2
        // JS Math.round: ties go up, same as floor(v + 0.5).
        fun clamp(value: Double, low: Double, high: Double): Double =
            floor(max(low, min(high, value)) * 10 + 0.5) / 10
        return FurniturePoint(clamp(point.x, edge, 100 - edge), clamp(point.y, minimumY, maximumY))
    }
}
